apollo-router-core/src/layers/deduplication.rs

```diff
@@ -64,12 +71,25 @@ where
                 }
                 None => {
                     let (tx, _rx) = broadcast::channel(1);
-                    locked_wait_map.insert(request.http_request.clone(), tx.clone());
+                    locked_wait_map
+                        .insert(request.http_request.clone(), Subscriber::new(tx.clone()));
                     drop(locked_wait_map);
 
                     let context = request.context.clone();
                     let http_request = request.http_request.clone();
-                    let res = service.ready_oneshot().await?.call(request).await;
+                    let mut s = match service.ready_oneshot().await {
```

Comment:
Is there not still a problem here if the future is dropped? The cleanup will never happen.

Reply:
The sequence of events for failure is at line 74 the subscriber is inserted into the `locked_wait_map`. The future is then dropped at line 80 at the `await`.
Neither the cleanup or the broadcast happens.

Reply:
yes, the future could be cancelled during the call to `ready_oneshot` too, and that would be caught in the same way, on the next similar request. Here I added an explicit match because in the case where `ready_oneshot` returns an error (so, not cancelled), we can clean up right away

Reply:
So if a similar request doesn't come in it leaks?

Reply:
Sorry, let me try rereading before answering.

Reply:
OK. I think I get it now.
There is an interesting thing that could occur if a request that is not the main calling thread is cancelled. The subsequent request will drop the the entry from the map even if the main calling thread is still in progress as the subscriber count will have decreased. Maybe we are not concerned about this.